pkg/kubelet/pod/mirror/manager.go

```diff
@@ -0,0 +1,372 @@
+	m := &basicManager{
+		staticPods:        make(map[string]*api.Pod),
+		mirrorPods:        make(map[string]*api.Pod),
+		invalidMirrorPods: make(map[string]*api.Pod),
+		apiserverClient:   apiserverClient,
+		podManager:        podManager,
+		// There shouldn't be too many mirror pod changes, set buffer size to 100
+		mirrorPodCreationChannel: make(chan *api.Pod, 100),
+		mirrorPodDeletionChannel: make(chan string, 100),
+	}
+	// Don't start mirror pod manager goroutine if we don't have a client. This will
+	// happen on the master, where the kubelet is responsible for bootstrapping the
+	// pods of the master components.
+	if apiserverClient == nil {
+		return m
+	}
+
+	glog.Info("Starting mirror pod manager syncing loop")
+	syncTicker := time.Tick(syncPeriod)
+	go util.Forever(func() {
+		select {
+		case pod := <-m.mirrorPodCreationChannel:
+			m.handleMirrorPodCreation(pod)
+		case pod := <-m.mirrorPodDeletionChannel:
+			m.handleMirrorPodDeletion(pod)
+		case <-syncTicker:
+			m.handleCleanup()
+		}
+	}, 0)
+	return m
+}
+
+func (m *basicManager) GetMirrorPod(podFullName string) *api.Pod {
+	m.mirrorPodsLock.RLock()
+	defer m.mirrorPodsLock.RUnlock()
+	return m.mirrorPods[podFullName]
+}
+
+func (m *basicManager) GetMirrorPods() []*api.Pod {
+	m.mirrorPodsLock.RLock()
+	defer m.mirrorPodsLock.RUnlock()
+	mirrorPods := []*api.Pod{}
+	for _, mirrorPod := range m.mirrorPods {
+		mirrorPods = append(mirrorPods, mirrorPod)
+	}
+	return mirrorPods
+}
+
+// No matter checking with staticPods or mirrorPods are both not accurate here when removing or
+// creating static pod and mirror pod.
+// However, it will return correct result when the system is stable, which is acceptable for status
+// manager. Now this method should only be used by status manager.
+func (m *basicManager) IsValidMirrorPod(mirrorPod *api.Pod) bool {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	// TODO: (random-liu) Use pod manager is more accurate.
+	podFullName := kubecontainer.GetPodFullName(mirrorPod)
+	staticPod, ok := m.staticPods[podFullName]
+	return ok && isMirrorPodOf(mirrorPod, staticPod)
+}
+
+func (m *basicManager) AddStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	podFullName := kubecontainer.GetPodFullName(staticPod)
+	if oldStaticPod, found := m.staticPods[podFullName]; found {
+		if isSameStaticPod(oldStaticPod, staticPod) {
+			// The static pod is the same with the old one, nothing to do.
+			return
+		}
+		glog.V(3).Infof("Static pod %q is changed, recreate the mirror pod", format.Pod(staticPod))
+	} else {
+		glog.V(4).Infof("New static pod %q, create the mirror pod", format.Pod(staticPod))
+	}
+
+	// The static pod is either new or changed, create a new mirror pod
+	m.staticPods[podFullName] = staticPod
+	m.createMirrorPod(staticPod)
+}
+
+func (m *basicManager) DeleteStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	podFullName := kubecontainer.GetPodFullName(staticPod)
```

Comment:
nit: I think you can add a `getKey(*api.Pod)` function to get the `podFullName`. You can then rename all `podFullName` in this file with `key` to keep code more readable and easier to change in the future.